rts/Sim/Features/FeatureHandler.cpp

```diff
@@ -203,6 +220,9 @@ void CFeatureHandler::Update()
 
 		updateFeatures.erase(iter, updateFeatures.end());
 	}
+	if ((gs->frameNum & 63) == 0) {
```

Comment:
> I set it to 62 because 2 secs felt right

How about

```suggestion
	/* Value arbitrary, represents a perf tradeoff:
	 * larger values decrease per-feature overhead over time for the usual state where features don't move vertically,
	 * smaller values decrease traceray overhead if features often enter and leave the stratosphere. */
	static constexpr auto UPDATE_RATE = 2 * GAME_SPEED;
	if (gs->frameNum % UPDATE_RATE == 0) {
```

Reply:
Shouldn't we also change the check above this for destroyed features for consistency too? Otherwise it feels a bit ad-hoc. I did it like this for internal consistency in the class tbh.

Also one concern here is everyone doing checks on frameNum % (n * GAME_SPEED), then all n multipliers of game speed would see perf hit with everyone queueing to do extra work there. So maybe want to add an offset here.

Right now all modules choose "slots" quite randomly resulting in random frames with increased load I guess.

Ideally maybe everyone should query for free/next update slots with some given period stepping so work is distributed among those, probably neither % or & standalone are really good in the end.

Maybe smth like:

```cpp
// example for doing specific slot in 32 (2⁵ frame groups).
static constexpr int PERIOD_STEPPING = 5;
static constexpr int UPDATE_MASK = std::pow(2, PERIOD_STEPPING)-1;
auto slot = engine->getNextSlot(PERIOD_STEPPING);

 if ((frameNum & UPDATE_MASK) % slot) // I think can be done as (frameNum+slot) & UPDATE_RATE
    // do work..
```

I think it could even be improved to sync between different periods. Lua could also use the mechanism, the only issue would be methods who want to do it right now and then with a specific period, but for those probably best would be to actually do it right now somehow, and then conform to the slot assigned by the engine.

Right now here, I think the consistency is important so the function uses the same method for both checks. Later maybe a method like above could be used to properly distribute load among processes.

Reply:
> Shouldn't we also change the check above this for destroyed features for consistency too? Otherwise it feels a bit ad-hoc. I did it like this for internal consistency in the class tbh.

Eventually yes. There are many `x & 31` left all over engine that should mostly become `x % GAME_SPEED` at some point, but some care needs to be taken for each to make sure there isn't a regression. In particular I am unsure if the one above is safe to change this way because running at a slightly slower rate than unit CommandAI slow updates may be on purpose to make sure CAI has an opportunity to process things.

> Right now here, I think the consistency is important so the function uses the same method for both checks.

Sure. I don't mind either way.

> everyone doing checks on frameNum % (n * GAME_SPEED), then all n multipliers of game speed would see perf hit with everyone queueing to do extra work there. So maybe want to add an offset here. Right now all modules choose "slots" quite randomly resulting in random frames with increased load I guess.

Sure. Most of this sort of periodic work is reasonably light though, I'd expect everything just has an offset of 0 but there aren't really spikes every second. Also there may be hidden assumptions so again it would take some care to make sure not to break things (e.g. handler A's periodic update must run right after handler B's and it's not made explicit anywhere, but it sort of happens to work because they just currently have the same offset and period - I remember something like this in resource income stat collection but there may be others).

> Lua could also use the mechanism, the only issue would be methods who want to do it right now and then with a specific period

Lua is a monolith from engine PoV so it is up to Lua to arrange work for individual wupgets, e.g. they could sign up dynamically to the wupgetHandler or use `Script.DelayByFrames`.

Reply:
> care needs to be taken for each to make sure there isn't a regression.

Indeed, any such rework would have to be done with extreme care to understand everything going on.

> Also there may be hidden assumptions so again it would take some care to make sure not to break things (e.g. handler A's periodic update must run right after handler B's and it's not made explicit anywhere, but it sort of happens to work because they just currently have the same offset and period

Current situation is fragile tho, because there might be such a situation happening and might be hard to notice, and any day someone may change some of the periods but then break any such assumptions, like if we decide to change both periods at featureHandler in this PR.